skorch/callbacks/training.py

```diff
@@ -761,3 +762,64 @@ def on_train_end(self, net, **kwargs):
         self.checkpoint_.save_model(net)
         self.checkpoint_._sink("Final checkpoint triggered", net.verbose)
         return self
+
+
+class InputShapeSetter(Callback):
+    """Sets the input dimension of the PyTorch module to the input dimension
+    of the training data. This can be of use when using a skorch model within
+    a sklearn pipeline and grid-searching feature transformers (or using
+    feature selection methods).
+
+    Basic usage:
+    >>> class MyModule(torch.nn.Module):
+    ...     def __init__(self, input_dim=1):
+    ...         super().__init__()
+    ...         self.layer = torch.nn.Linear(input_dim, 3)
+    ... # ...
+    >>> X1 = np.zeros(100, 5)
+    >>> X2 = np.zeros(100, 3)
+    >>> y = np.zeros(100)
+    >>> net = NeuralNetClassifier(MyModule, callbacks=[InputShapeSetter()])
+    >>> net.fit(X1, y) # sets self.module_ = module(input_dim=5)
+    >>> net.fit(X2, y) # sets self.module_ = module(input_dim=3)
```

Comment:
Not sure if it's better, just a suggestion.

```suggestion
    >>> net.fit(X1, y)  # self.module_.layer.in_features == 5
    >>> net.fit(X2, y)  # self.module_.layer.in_features == 3
```